engine/src/main/java/org/terasology/world/chunks/blockdata/ExtraBlockDataManager.java

```diff
@@ -0,0 +1,320 @@
+     */
+    private static ArrayList<ArrayList<String>> findCliqueCover(Graph graph) {
+        return findCliqueCover(graph, Integer.MAX_VALUE, "");
+    }
+    
+    private static ArrayList<ArrayList<String>> findCliqueCover(Graph graph, int bestSize, String tabs) {
+        verboseLog(tabs + "findCliqueCover up to " + bestSize + ", " + graph.toString());
+        for (int i = 0; i < graph.verts.length; i++) {
+            if (i >= bestSize - 1) {
+                verboseLog(tabs + "giving up");
+                return null;
+            }
+            String v0 = graph.verts[i];
+            if (!graph.edges.get(v0).isEmpty()) {
+                verboseLog(tabs + "Selected vertex " + v0);
+                String v1 = graph.edges.get(v0).iterator().next();
+                ArrayList<ArrayList<String>> bestCover0 = findCliqueCover(graph.ntract(v0, v1), bestSize, tabs + "----");
+                int bestSize0 = bestCover0 == null ? bestSize : bestCover0.size();
+                graph.removeEdge(v0, v1);
+                ArrayList<ArrayList<String>> bestCover1 = findCliqueCover(graph, bestSize0, tabs + "    ");
+                graph.addEdge(v0, v1);
+                if (bestCover1 != null) {
+                    return bestCover1;
+                } else {
+                    if (bestCover0 != null) {
+                        bestCover0.get(i).add(v1);
+                    }
+                    return bestCover0;
+                }
+            }
+        }
+        verboseLog(tabs + "done, " + graph.verts.length);
+        ArrayList<ArrayList<String>> bestCover = new ArrayList<>();
+        for (int i = 0; i < graph.verts.length; i++) {
+            ArrayList<String> singleton = new ArrayList<>();
+            singleton.add(graph.verts[i]);
+            bestCover.add(singleton);
+        }
+        return bestCover;
+    }
+    
+    // Log something, but only when this class is being tested.
+    private static void verboseLog(String string) {
+        //logger.info(string);
+    }
+    
+    /**
+     * Get the numerical index associated with the extra data field name.
+     * This numerical index is needed for most extra-data access methods.
+     */
+    public int getSlotNumber(String name) {
+        Integer index = slots.get(name);
+        if (index == null) {
+            throw new IllegalArgumentException("Extra-data name not registered: " + name);
+        }
+        return index;
+    }
+    
+    public TeraArray[] makeDataArrays(int sizeX, int sizeY, int sizeZ) {
+        TeraArray[] extraData = new TeraArray[slotFactories.length];
+        for (int i = 0; i < extraData.length; i++) {
+            extraData[i] = slotFactories[i].create(sizeX, sizeY, sizeZ);
+        }
+        return extraData;
+    }
+    
+    /**
+     * Undirected graphs with string-labelled vertices.
+     * Used to represent the overlaps between requested extra-data fields.
+     */
+    private static class Graph {
```

Comment:
Hmm, not sure about the design of this class.
It looks big and complex, but it doesn't provide any encapsulation.
The constructors are also error prone, since they make shallow copies of the params.
If you just want Graph to be a small convenience class with low overhead, consider something like this:

```
// This screams: I am just data. Responsibility of valid contents lies with the users.
private static class Graph final {
   public String[] verts;
   public Map<String, Set<String>> edges;
}

private static Graph ntract(Graph g, String s0, String s1) {
   ...
}
```

Reply:
It seems tidier to me to have all the purely graph-related stuff clearly grouped together. It was my intention to have the Graph class enforcing its own invariants, but I suppose you're correct that it can't entirely guarantee that in its current form. I think I'd rather go in the direction of strengthening its encapsulation than weakening it (in the sense that although your suggested change doesn't actually provide any more ways to break encapsulation, it encourages doing so (this isn't intended as a complaint at your suggestion, just a clarification)).

Reply:
You can go either way, full encapsulation or just raw data.